src/merkle_tree/mod.rs

```diff
@@ -1,209 +1,302 @@
-use crate::{crh::FixedLengthCRH, Vec};
-use ark_ff::bytes::ToBytes;
-use ark_std::fmt;
+#![allow(unused)] // temporary
 
+use crate::crh::TwoToOneCRH;
+use crate::CRH;
+use ark_ff::ToBytes;
+use ark_std::vec::Vec;
 #[cfg(feature = "r1cs")]
 pub mod constraints;
 
 pub trait Config {
-    const HEIGHT: usize;
-    type H: FixedLengthCRH;
+    type LeafHash: CRH;
+    type TwoToOneHash: TwoToOneCRH;
+    /// Determine the upper bound of output size of leaf hash.
+    fn leaf_hash_output_size_upper_bound() -> usize;
+    /// Determine the upper bound of output size of two-to-one hash.
+    fn two_to_one_hash_output_size_upper_bound() -> usize;
 }
 
-/// Stores the hashes of a particular path (in order) from leaf to root.
-/// Our path `is_left_child()` if the boolean in `path` is true.
-#[derive(Derivative)]
-#[derivative(
-    Clone(bound = "P: Config"),
-    Debug(bound = "P: Config, <P::H as FixedLengthCRH>::Output: fmt::Debug")
-)]
+pub type TwoToOneDigest<P> = <<P as Config>::TwoToOneHash as TwoToOneCRH>::Output;
+pub type LeafDigest<P> = <<P as Config>::LeafHash as CRH>::Output;
+pub type TwoToOneParam<P> = <<P as Config>::TwoToOneHash as TwoToOneCRH>::Parameters;
+pub type LeafParam<P> = <<P as Config>::LeafHash as CRH>::Parameters;
+
+/// Stores the hashes of a particular path (in order) from root to leaf.
+/// For example:
+/// ```tree_diagram
+///          A
+///        /  \
+///       B   C
+///     / \  / \
+///    D  E F  H
+///   .. / \ ....
+///     I  J
+/// ```
+///  Suppose we want to prove I, then `leaf_and_sibling_hash` is (I, J), `non_leaf_and_sibling_hash` is
+///  [(B,C),(D,E)]
 pub struct Path<P: Config> {
-    pub(crate) path: Vec<(Digest<P>, Digest<P>)>,
+    pub(crate) leaf_and_sibling_hash: (LeafDigest<P>, LeafDigest<P>),
+    /// The hash path from higher layer to lower layer.
+    pub(crate) non_leaf_and_sibling_hash_path: Vec<(TwoToOneDigest<P>, TwoToOneDigest<P>)>,
```

Comment:
Can we change these data structures to only include the authentication path? (All the sibling hashes, not the hashes that the verifier computes on its own)

Reply:
Makes sense. Just realized that we can compute node path from leaf... I will change it

Reply:
Awesome, thanks! The fractal constraints merkle tree has constraints code for that if it helps

Reply:
updated